test/testdata/infer/no_short_circuit_type_constraint_nested.rb

```diff
@@ -0,0 +1,84 @@
+# typed: true
+
+# The problem demonstrated in this test is very similar to the one in
+# no_short_circuit_type_constraint.rb. The core problem boils down to Sorbet's
+# type constraints accumulating greedily.
+#
+# There is more context in that test. For the time being, it's not _wrong_ for
+# Sorbet to infer this type, it's just not quite as narrow as you'd want it to
+# be.
+#
+# It's also worth noting that sometimes users don't have control over how their
+# nested types are created--the order may be determined by the order that
+# Sorbet lubs basic block branches and glbs type tests. So sometimes the bad
+# case below can happen because of no fault of the user.
+
+module M1; end
+module M2; end
+
+module A; end
+module B; end
+
+module Main
+  extend T::Sig
+
+  # Type signature indicates that we're trying to remove the `B` from the input.
+  sig do
+    type_parameters(:U)
+      .params(x: T.any(T.type_parameter(:U), B))
+      .returns(T.type_parameter(:U))
+  end
+  def self.remove_the_b(x)
+    case x
+    when B then raise
+    else x
+    end
+  end
+
+  sig do
+    type_parameters(:U)
+      .params(x: T.all(M1, T.any(A, B), M2))
+      .void
+  end
+  def self.test1(x)
+    # Sorbet gets this type because when faced with a type like
+    #
+    #   X & Y  <:  A | B
+    #
+    # It needs to decide which comparison to do first: split the AndType:
+    #   X & Y <: A   ||   X & Y <: B
+    # or split the OrType:
+    #   X <: A | B   ||   Y <: A | B
```

Comment:
> One of these expressions should be && instead of ||, surely?

Counterintuitively, that's not the case.

- If either of `X & Y <: A` or `X & Y <: B` holds, then certainly also `X & Y <: A | B`, because `A <: A | B` and `B <: A | B`
- If either of `X <: A | B` or ` Y <: A | B` holds, then certainly also `X & Y <: A | B`, because `X & Y <: X` and `X & Y <: Y`.

Reply:
> "splitting the AndType" suggests to me that we should be doing X <: A | B || Y <: A | B, as the AndType shouldn't appear in the resulting expression. Similarly for the OrType split. Or did you mean to talk about distributing, as with the comment on the below function?

I think i just swapped `AndType` and `OrType`. It now reads `split the OrType: ... or split the AndType:`